src/main/java/net/neoforged/neoforge/fluids/CauldronFluidContent.java

```diff
@@ -0,0 +1,169 @@
+     */
+    public final int totalAmount;
+    /**
+     * Maximum level for {@link #levelProperty}. {@code 1} if {@code levelProperty} is null, otherwise a number {@code >= 1}.
+     * The minimum level is always 1.
+     */
+    public final int maxLevel;
+    /**
+     * Property storing the level of the cauldron. If it's {@code null}, only one level is possible.
+     */
+    @Nullable
+    public final IntegerProperty levelProperty;
+
+    /**
+     * Return the current level of the cauldron given its block state, or 0 if it's an empty cauldron.
+     */
+    public int currentLevel(BlockState state) {
+        if (fluid == Fluids.EMPTY) {
+            return 0;
+        } else if (levelProperty == null) {
+            return 1;
+        } else {
+            return state.getValue(levelProperty);
+        }
+    }
+
+    private CauldronFluidContent(Block block, Fluid fluid, int totalAmount, int maxLevel, @Nullable IntegerProperty levelProperty) {
+        this.block = block;
+        this.fluid = fluid;
+        this.totalAmount = totalAmount;
+        this.maxLevel = maxLevel;
+        this.levelProperty = levelProperty;
+    }
+
+    private static final Map<Block, CauldronFluidContent> BLOCK_TO_CAULDRON = new IdentityHashMap<>();
+    private static final Map<Fluid, CauldronFluidContent> FLUID_TO_CAULDRON = new IdentityHashMap<>();
+
+    /**
+     * Get the cauldron fluid content for a cauldron block, or {@code null} if none was registered (yet).
+     */
+    @Nullable
+    public static CauldronFluidContent getForBlock(Block block) {
+        return BLOCK_TO_CAULDRON.get(block);
+    }
+
+    /**
+     * Get the cauldron fluid content for a fluid, or {@code null} if no cauldron was registered for that fluid (yet).
+     */
+    @Nullable
+    public static CauldronFluidContent getForFluid(Fluid fluid) {
+        return FLUID_TO_CAULDRON.get(fluid);
+    }
+
+    @ApiStatus.Internal
+    public static void init() {
+        var registerEvent = new RegisterCauldronFluidContentEvent();
+        // Vanilla registrations
+        registerEvent.register(Blocks.CAULDRON, Fluids.EMPTY, 1000, null);
+        registerEvent.register(Blocks.WATER_CAULDRON, Fluids.WATER, 1000, LayeredCauldronBlock.LEVEL);
+        registerEvent.register(Blocks.LAVA_CAULDRON, Fluids.LAVA, 1000, null);
+        // Modded registrations
+        ModLoader.get().postEvent(registerEvent);
+    }
+
+    /**
+     * Do not try to call, use the {@link RegisterCauldronFluidContentEvent} event instead.
+     */
+    static void register(Block block, Fluid fluid, int totalAmount, @Nullable IntegerProperty levelProperty) {
+        if (BLOCK_TO_CAULDRON.get(block) != null) {
+            throw new IllegalArgumentException("Duplicate cauldron registration for block %s.".formatted(block));
+        }
+        if (FLUID_TO_CAULDRON.get(fluid) != null) {
+            throw new IllegalArgumentException("Duplicate cauldron registration for fluid %s.".formatted(fluid));
+        }
```

Comment:
How much do we care about checking these immediately? Because in general it shouldn't happen and modders should test their code (at least by launching the game), so whether it happens at the top of this method, or at the bottom by checking if the `put` calls had the values at previously being `null` shouldn't matter? And then it allows bypassing an extra lookup for each of these.

Reply:
I like to have preconditions at the top of the method. And it feels a bit dirty to potentially mutate the first map then crash on the second.